### PR TITLE
SI-7247, deprecate NotNull.

### PR DESCRIPTION
**Pull request by paulp**
Brought the steamroller to bear on NotNull; fired up the Zamboni afterward to smooth things over.

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -462,12 +461,9 @@ trait Types
...
     def boundSyms: immutable.Set[Symbol] = emptySymbolSet
 
-    /** Mixin a NotNull trait unless type already has one
-     *  ...if the option is given, since it is causing typing bugs.
+    /** Obsolete, here for backward compatibility.
      */
-    def notNull: Type =
-      if (!settings.Ynotnull.value || isNotNull || phase.erasedTypes) this
-      else NotNullType(this)
+    @deprecated("This method will be removed", "2.11.0") def notNull: Type = this
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala]: we're in reflect.internal -- we can probably get rid of this as well

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -2347,8 +2333,7 @@ trait Types
     override def typeSymbol       = sym
     override def typeSymbolDirect = sym
 
-    override def isNotNull =
-      sym.isModuleClass || sym == NothingClass || (sym isNonBottomSubClass NotNullClass) || super.isNotNull
+    override def isNotNull = sym == NothingClass || super.isNotNull
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala]: is this still referenced?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 3]
```diff
@@ -4080,8 +4065,7 @@ trait Types
 
   protected[internal] def containsNull(sym: Symbol): Boolean =
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala]: also unreferenced, I think

**Comment by adriaanm** [comment]
Great. I almost finished cobbling together my own PR when I checked email and saw this. This LBTM (think "better", not "lobotomy") once we prune those last unused things, unless they are used?

File: src/reflect/scala/reflect/internal/Types.scala (hunk 2 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala]: isNotNull is not used much if at all, but if you read my commit messages (github seems to discourage this, lame) you will see it was intentional.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 3 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala]: containsNull is used in subtyping - Null <:< X if containsNull(X).

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala]: Would be glad to dump it, but I'm always breaking the IDE and I figured a one line identity function was cheap insurance.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 2 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala]: You know what: IAGNI. You are right, enjoy this moment. I'll remove it.

**Comment by paulp** [comment]
Will rebase/remove/reopen.

**Comment by paulp** [comment]
Okay @adriaanm this should now be worthy of your attention.

**Comment by adriaanm** [comment]
I hope so.

**Comment by paulp** [comment]
@adriaanm, perhaps it's time for YOU to try to be worthy of IT. Too busy making smoothies, I bet.

**Comment by adriaanm** [comment]
LGTM

**Comment by adriaanm** [comment]
Great JIRA is "temporarily unavailable" (I wish it was permanent). Ticket still needs to be closed.

**Comment by lrytz** [comment]
i started it again

**Comment by cchantep** [comment]
If I properly get it, if a value whose type is currently marked with this trait is given a `null`, a compilation error is raised (scala version = 2.10.4-RC2, using SBT 0.13.1):

```
type mismatch;
[error]  found   : Null(null)
[error]  required: ANotNullType
```

What will replacement for such case, either for scala-lib type (e.g. `Int`, `Float`, ...) or those outside?

**Comment by adriaanm** [comment]
NotNull only worked in trivial cases and it proved hard to extend it beyond
them in a useful way, so we'd rather not have it at all.

**Comment by cchantep** [comment]
I can understand, but that's a quite radical pov.

**Pull request merged**